File: srclib/v3/R/arguments.R
# arguments function

arguments <- function(name, dir, mode) {
  name <<- sub("^\\s+", "", sub("\\s+$", "", name))  # name
  if (identical(name, "")) msg <<- "Missing name."
  else {
    if (!grepl("^[[:alnum:]][[:alnum:]_ -]*$", name)) msg <<- "Invalid name."
  }
  if (identical(msg, "")) {  # path
    dir <- sub("^\\s+", "", sub("\\s+$", "", dir))
    if (identical(dir, "")) path <<- name  # no directory specified, use just name
    else {  # specified directory
      if (grepl('^"([^"]*)"$', dir)) {  # double quotes
        dir <- sub('^"([^"]*)"$', "\\1", dir)
        dir <- sub("^\\s+(.*)", "\\1", sub("\\s+$", "", dir))
      }
      else if (grepl("^'([^']*)'$", dir)) {  # single quotes
        dir <- sub("^'([^']*)'$", "\\1", dir)
        dir <- sub("^\\s+(.*)", "\\1", sub("\\s+$", "", dir))
      }
      if (!grepl("^[~\\.\\\\/:[:alnum:]_ -]+$", dir))  msg <<- "Invalid directory."
      else path <<- file.path(dir, name, fsep=.Platform$file.sep)
     }
  }
  if (identical(msg, "")) {  # test
    test <<- sub("^\\s+", "", sub("\\s+$", "", mode))
    if (!identical(test, "")) {
      if (!grepl("^(debug|test)$", test, ignore.case=TRUE)) msg <<- "Invalid test."
    }
  }
  if (identical(msg, "")) {  # get the setup
    setup <<- ""
    set <- paste(path, "set", sep=".")
    if (file.exists(set)) setup <<- readChar(set, file.info(set)$size)
    if (identical(setup, "")) msg <<- "Missing setup."
  }
  if (identical(msg, "")) {  # datain
    if (grepl("(^|;)\\s*datain\\s+[^;]+;", setup, ignore.case=TRUE)) {
      datain <<- sub("^(.*;)*\\s*datain\\s+([^;]+);.*$", "\\2", setup, ignore.case=TRUE)  # files
      datain <<- sub("^\\s+", "", sub("\\s+$", "", datain))  # trim
      if (identical(proc, "combine")) {  # combine
        if (!grepl("^[\"'~\\.\\\\/:[:alnum:]_ -]+$", datain))  msg <- "Invalid datain."
      }
      else {  # other prodedures
        datain <<- gsub('\\s+"([^"]*)"', "", datain)  # remove double quoted names after the first
        datain <<- gsub("\\s+'([^']*)'", "", datain)  # ditto single quoted
        datain <<- gsub("\\s+([^\\s]*)", "", datain)  # ditto unquoted
        if (grepl('^"[^"]*"$', datain)) datain <<- sub('^"\\s*(.*)\\s*"$', "\\1", datain)  # remove double quotes
        else if (grepl("^'[^']*'$", datain)) datain <<- sub("^'\\s*(.*)\\s*'$", "\\1", datain)  # remove single quotes
        if (!grepl("^[~\\.\\\\/:[:alnum:]_ -]+$", datain))  msg <<- "Invalid datain."  # check name
      }
    }
    else msg <<- "Missing datain."
  }
  if (identical(msg, "")) {  # dataout
    if (grepl("(^|;)\\s*dataout\\s+[^;]+;", setup, ignore.case=TRUE)) {
      dataout <<- sub("^(.*;)*\\s*dataout\\s+([^;]+);.*$", "\\2", setup, ignore.case=TRUE)  # file
      if (grepl("\\s+(all|con|concat|concatenate)\\s*$", dataout, ignore.case=TRUE)) {  # remove all or concatenate
        dataout <<- sub("\\s+(all|con|concat|concatenate)\\s*$", "", dataout, ignore.case=TRUE)
      }
      dataout <<- sub("^\\s+", "", sub("\\s+$", "", dataout))  # trim
      if (grepl('^"[^"]*"$', dataout)) dataout <<- sub('^"\\s*(.*)\\s*"$', "\\1", dataout)  # remove double quotes
      else if (grepl("^'[^']*'$", dataout)) dataout <<- sub("^'\\s*(.*)\\s*'$", "\\1", dataout)  # remove single quotes
      if (!grepl("^[~\\.\\\\/:[:alnum:]_ -]+$", dataout))  msg <<- "Invalid dataout."  # check name
    }
    else if (identical(proc, "combine")) msg <<- "Missing dataout."
  }
}

File: srclib/v3/spss/srclib/arguments.py
# arguments function

import os, re, spssaux, sys

def arguments(name, dir, mode, proc):
  msg = None
  if not name:  # name
    msg = "Missing run name."
  else:
    name = name.strip()
    m = re.match(r"^[\w -]+$", name)  # check name
    if not m:
      msg = "Invalid name."
  if not msg:
    cwd = spssaux.getShow("DIRECTORY")  # spss current working directory
    if not dir:
      path = os.path.join(cwd, name)  # spss current working directory
    else:  # specified directory
      dir = dir.strip()
      m = re.match(r'^"([^"]+)"$', dir)  # remove double quotes
      if m:
        dir = m.group(1).strip()
      else:
        m = re.match(r"^'([^']+)'$", dir) # remove single quotes
        if m:
          dir = m.group(1).strip()
      m = re.match(r"^[~\.\\/:\w -]+$", dir)  # check dir
      if not m:
        msg = "Invalid directory."
      if not msg:
        path = os.path.join(os.path.normpath(dir), name)
  if not msg:
    if mode:
      mode = mode.strip()
      if not re.match(r"^(debug|test)$", mode, flags=re.IGNORECASE):  # mode
        msg = "Mode error."
  if not msg:
    setup = ""  # get the setup
    with open(path + ".set", "r") as f:
      for line in f:
        setup += line
    if not setup:
      msg = "Missing setup."
  if not msg:
    m = re.search(r"(^|;)\s*datain\s+([^;]+);", setup, flags=re.IGNORECASE)  # datain
    if m:
      datain = m.group(2).strip()
      if proc != "combine":
        m = re.match(r'^"([^"]+)"', datain)  # remove double quotes
        if m:
          datain = m.group(1).strip()
        else:
          m = re.match(r"^'([^']+)'", datain)  # remove single quotes
          if m:
            datain = m.group(1).strip()
          else:
            m = re.match(r"^([^ ]+)", datain)  # remove subsequent file names
            if m:
              datain = m.group(1).strip()
        m = re.match(r"^[~\.\\/:\w -]+$", datain)  # check name
        if not m:
          msg = "Invalid datain."
    else:
      msg = "Missing datain."
  if not msg:
    m = re.search(r"(^|;)\s*dataout\s+([^;]+);", setup, flags=re.IGNORECASE)  # dataout
    if m:
      dataout = m.group(2).strip()
      m = re.match(r"^(.+)\s+(all|con|concat|concatenate)$", dataout, flags=re.IGNORECASE)  # rmove all or concatenate
      if m:
        dataout = m.group(1).strip()
      m = re.match(r'^"([^"]+)"$', dataout)  # remove double quotes
      if m:
        dataout = m.group(1).strip()
      else:
        m = re.match(r"^'([^']+)'$", dataout)  # remove single quotes
        if m:
          dataout = m.group(1).strip()
      m = re.match(r"^[~\.\\/:\w -]+$", dataout)  # check name
      if not m:
        msg = "Invalid dataout."
      if not msg:
        if sys.platform.startswith("win"):  # windows
          m = re.match(r"^([a-zA-Z]:|[\\/])", dataout)
        else:  # linux
          m = re.match("^(~|/)", dataout)
        if not m:  # not full path
          dataout = os.path.join(cwd, dataout)  # prefix the current working directory
        dataout = os.path.normpath(dataout)  # normalize the path
    else:
      dataout = None  # no dataout
      if proc == "combine":
        msg = "Missing dataout."
  if msg:
    print msg  # print the error message
    return None
  return [path, datain, dataout, setup]

